daemon/graphdriver/aufs/aufs.go

```diff
@@ -308,11 +308,17 @@ func (a *Driver) Put(id string) error {
 
 // Diff produces an archive of the changes between the specified
 // layer and its parent layer which may be "".
-func (a *Driver) Diff(id, parent string) (archive.Archive, error) {
+func (a *Driver) Diff(id, parent string, excludes []string) (archive.Archive, error) {
+	// archive.TarWithOptions uses relative path to compare excludes.
+	// So we have to strip the leading '/'.
+	newExcludes := make([]string, len(excludes))
+	for i, ex := range excludes {
+		newExcludes[i] = strings.TrimLeft(ex, "/")
+	}
```

Comment:
@icecrime In this line https://github.com/docker/docker/blob/master/pkg/archive/archive.go#L435 the srcPath is converted to a relative path (relative to the diff path of the container).
If the leading '/' was not stripped before calling `archive.TarWithOptions`,  when the user run `docker commit --exclude=/path/file.txt cid img`, `fileutils.OptimizedMatches` won't mach the pattern and the file won't be excluded.